### PR TITLE
Support Untracked changes only status for the Commit Icon.

### LINKED ISSUE #3887: Feature Req.: Commit button should indicate if file in repos. changed
**Do you want to request a *feature* or report a *bug*?**
Feature

**What is the current behavior?**
If there are files in working copy which are not in repository (untracked) and no change in index, the commit button is orange regardless if a tracked file was changed or not.

**What is the expected behavior?**

The commit button (or alternatively the stash button) should look differently if no tracked file is modified compared to if tracked files are modified. This is helpful to know if checking out a different branch or hard resetting current branch to a particular commit can be done without hassle and loosing changes in the working copy.

**Which versions of GitExtensions, GIT and Windows are affected by this issue?**

GitExtensions 2.50.01

### PR DESCRIPTION
**Pull request by Drake103**
Fixes #3887.

Changes proposed in this pull request:
 - Add new Icon for the `UntrackedChangesOnly` status
 - Extract icon selection logic to `CommitButtonStatusProvider`
 - Introduce `ICommitButtonStatus` and a separate class for each status.
 
Screenshots before and after (if PR changes UI):
before:
image
after (pale green):
image

How did I test this code:
 - Cover `CommitButtonStatusProvider` with unit tests
 - Manual testing

Has been tested on (remove any that don't apply):
 - GIT 2.10 and above
 - Windows 10

### CODE REVIEW

**Comment by RussKie** [comment]
I feel that with the last commit you over-complicated the design.

Each status is just that - a single piece of information represented by an image. 
From the point of interface driven development DTOs/POCOs do not need interfaces because they have no behaviors. They may warrant an interface for abstraction reasons, but not in this case.

So a change which spanned 4 files now spans across 15 (slightly less if we move interfaces to implementations but still)...

I like your implementation for `CommitIconProvider` - I think it is sufficient.

**Comment by Drake103** [comment]
Main reason: I will have to test `CommitIconProvider` for concete Bitmaps (I don't like testing for concrete Resources).
Each status represents a Commit button apperance. My point is that Commit button states are presented not only with Image but also with Labels (Commit vs Commit (5)). Different images are provided using dynamic dispatch, this seems a behavior for me.
I agree, that providing only `Image` is really over-complication. Should I revert it?

**Comment by RussKie** [comment]
> I will have to test CommitIconProvider for concete Bitmaps

I have no problems with that, these bitmaps are already compiled into the assembly anyway.

> Different images are provided using dynamic dispatch, this seems a behavior for me.

That's the behavior of the dispatcher not the images per se.

You could probably change the `CommitIconProvider` to `CommitStatusProvider` which returns `CommitStatus` poco object that could contain the status image and status format.
Not sure how this useful this may be, up to you if you wish to explore it further.

I also do not see any reason to wrap the images into properties - `Properties.Resources.xxx` are already static, extra layer just adds extra CPU cycles. 

```
 +        private static Bitmap IconClean
 +        {
 +            get { return Properties.Resources.IconClean; }
 +        }
```

> I agree, that providing only Image is really over-complication. Should I revert it?

Yes, less code the better - less code to maintain, less bugs to fix :)

**Comment by Drake103** [comment]
I have dropped the last commit and changed tests a bit to check for Bitmaps. Unfortunately each Bitmap resource `get` creates new instance of `Bitmap` class. Default `Equals` returns `false`, since they are reference types. I decided to share Bitmaps using public static readonly fields to avoid writing custom equaler.

**Comment by Drake103** [comment]
Fixed small glich with CommitButton:
before:

**Comment by RussKie** [comment]
> Unfortunately each Bitmap resource get creates new instance of Bitmap class. 

I'm learning something new each day (or remembering things I've long forgotten) - Winforms image management is shocking...
https://theartofdev.com/2013/06/03/optimizing-images-usage-in-net-part-2/

I guess it is an area for us to improve in the future.

File: GitUI/UserControls/ToolStripClasses/CommitIconProvider.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+        public static readonly Bitmap IconDirty = Properties.Resources.IconDirty;
+
+        public static readonly Bitmap IconDirtySubmodules = Properties.Resources.IconDirtySubmodules;
+
+        public static readonly Bitmap IconStaged = Properties.Resources.IconStaged;
+
+        public static readonly Bitmap IconMixed = Properties.Resources.IconMixed;
+
+        public static readonly Bitmap IconUntrackedOnly = Properties.Resources.IconUntrackedOnly;
```
Comment (RussKie) [inline GitUI/UserControls/ToolStripClasses/CommitIconProvider.cs]: these are probably best made internal

**Comment by jbialobr** [comment]
Why did you choose green color? For me, green should be reserved for 'no changes, all clear and ok status'.

**Comment by Drake103** [comment]
I used pale green, because for me it was like "no tracked changes/almost no changes" (e.g. can checkout without resetting local directory). Another color I consider is violet, since it will be unique. Tell me the color, I will change it.

**Comment by jbialobr** [comment]
Choose any unique color you like.

**Comment by jbialobr** [comment]
Thank you.

**Pull request merged**